### ISSUE TITLE
Are VXL exception functions just aborting?

### ISSUE THREAD

**Issue by DVigneault** [vxl/vxl#206]
### TL;DR

According to the [implementation file](https://github.com/vxl/vxl/blob/647386048f3f71ed6d9e7d67de5d11dab5ddacfc/core/vnl/vnl_error.cxx) documentation (there is no documentation in the header), the function `vnl_error_vector_index` should "[r]aise exception for invalid index".  However, my attempts to `throw` and `catch` with this function resulted in program termination.  After drilling down into the preprocessor definitions, it looks as though the function is just calling `std::abort()`, and no exception is being raised.  My questions are:

1.  Am I correct that `vnl_error_vector_index`, and all the other functions in that file, are just calling `std::abort()`?
2.  If so, what should be the actual behavior?  It seems to me that if the implementation is correct, the documentation needs to be updated, or vice versa.

### The Gory Details

Let's say I want to catch an exception thrown in the standard library.  I can do it like this:

```
  std::vector<int> std_vec;
  try {std_vec.at(25);}
  catch (...) { vcl_cout << "This gets caught." << vcl_endl; }
```
I thought perhaps I could do the same for a vxl exception, but the exception does not get caught:
```
  try { vnl_error_vector_index("get", 25); }  // Raise exception
  catch (...) { vcl_cout << "Not caught." << vcl_endl; }
```
After poking around in the code for a while, I found that vxl actually defines its own try/catch structure, but unfortunately that doesn't catch it either:
```
  vcl_try { vnl_error_vector_index("get", 25); }  // Raise exception
  vcl_catch_all { vcl_cout << "Not caught." << vcl_endl; }
```
Both of the vxl examples give an error and terminate execution.
```
vnl_error_vector_index:get: Invalid value 25 specified for index.
```
So I poked around some more and found the implementation of `vxl_error_vector_index` [here](https://github.com/vxl/vxl/blob/647386048f3f71ed6d9e7d67de5d11dab5ddacfc/core/vnl/vnl_error.cxx):

```
//: Raise exception for invalid index
void vnl_error_vector_index (char const* fcn, int index)
{
  //RAISE Error, SYM(vnl_error_vector), SYM(Invalid_Index),
  vcl_cerr << "vnl_error_vector_index:" << fcn
           << ": Invalid value " << index << " specified for index.\n";
  vcl_abort();
}
```
A little more investigating into `vcl_abort()` seems to show that no exception is actually being thrown.  From [`vcl/generic/vcl_cstdlib.h`](https://github.com/vxl/vxl/blob/647386048f3f71ed6d9e7d67de5d11dab5ddacfc/vcl/generic/vcl_cstdlib.h):
```
// abort
#ifndef vcl_abort
#define vcl_abort vcl_generic_cstdlib_STD :: abort
#endif
```
And from [`vcl/iso/vcl_stdlib.h`](https://github.com/vxl/vxl/blob/647386048f3f71ed6d9e7d67de5d11dab5ddacfc/vcl/iso/vcl_cstdlib.h):
```
#ifdef vcl_generic_cstdlib_STD
  ** error **
#else
# define vcl_generic_cstdlib_STD std
#endif
```
So all in all `vcl_abort()` seems to resolve to `std::abort()`.

**Comment by DVigneault** [comment]
This issue was addressed here https://github.com/vxl/vxl/pull/207.
